ibis/common/caching.py

```diff
@@ -47,56 +49,50 @@ def __init__(
         generate_name: Callable[[], str],
         key: Callable[[Any], Any],
     ) -> None:
-        self.cache = bidict()
-        # Somehow mypy needs a type hint here
-        self.refs: Counter = Counter()
         self.populate = populate
         self.lookup = lookup
         self.finalize = finalize
-        # Somehow mypy needs a type hint here
-        self.names: defaultdict = defaultdict(generate_name)
+        self.generate_name = generate_name
         self.key = key or (lambda x: x)
 
+        self.cache: dict[Any, CacheEntry] = dict()
```

Comment:
Not to relitigate this whole PR but would a `WeakValueDictionary` also work here instead of explicit finalizers?

Not sure if `WeakValueDictionary` lets you control what happens on weakref finalization, but if it doesn't that I suppose that would be a non-starter.

Reply:
Seems like the finalizer behavior of the values in a `WeakValueDictionary` can't easily be customized.

Reply:
I actually had one originally and look into the implementation whether subclassing one could easily replace what i am doing here. I did not see a good way to do it more easily.